src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs

```diff
@@ -46,54 +68,225 @@ internal SafeWaitHandle[] GetWaitedSafeWaitHandleArray(int requiredCapacity)
 
-                t_currentThread = currentThread = new RuntimeThread();
-                Debug.Assert(currentThread._threadState == (int)ThreadState.Unstarted);
-                currentThread._threadState = (int)(ThreadState.Running | ThreadState.Background);
-                return currentThread;
+        public bool IsThreadPoolThread
+        {
+            get
+            {
+                if (IsDead)
+                {
+                    throw new ThreadStateException(SR.ThreadState_Dead_State);
+                }
+                return GetThreadStateBit(ThreadPoolThread);
             }
         }
 
-        public bool IsAlive { get { throw null; } }
-        public bool IsBackground { get { throw null; } set { throw null; } }
-        public bool IsThreadPoolThread { get { throw null; } }
-        public int ManagedThreadId { get { throw null; } }
-        public string Name { get { throw null; } set { throw null; } }
-        public ThreadPriority Priority { get { throw null; } set { throw null; } }
+        public int ManagedThreadId => _managedThreadId.Id;
 
-        public ThreadState ThreadState => (ThreadState)Volatile.Read(ref _threadState);
+        public string Name
+        {
+            get
+            {
+                return _name;
+            }
+            set
+            {
+                using (LockHolder.Hold(_lock))
+                {
+                    if (_name != null)
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_WriteOnce);
+                    }
+                    _name = value;
+                    // TODO: Inform the debugger and the profiler
+                }
+            }
+        }
+
+        public ThreadPriority Priority
+        {
+            get
+            {
+                if (IsDead)
+                {
+                    throw new ThreadStateException(SR.ThreadState_Dead_Priority);
+                }
+                return _priority;
+            }
+            set
+            {
+                if ((value < ThreadPriority.Lowest) || (ThreadPriority.Highest < value))
+                {
+                    throw new ArgumentOutOfRangeException(SR.Argument_InvalidFlag);
+                }
+                if (IsDead)
+                {
+                    throw new ThreadStateException(SR.ThreadState_Dead_Priority);
+                }
+
+                // Possible race condition with starting this thread
+                using (LockHolder.Hold(_lock))
+                {
+                    if (!SetPriority(value))
+                    {
+                        throw new ThreadStateException(SR.ThreadState_SetPriorityFailed);
+                    }
+                    _priority = value;
+                }
+            }
+        }
+
+        public ThreadState ThreadState => GetThreadState();
+
+        private ThreadState GetThreadState()
+        {
+            int state = Volatile.Read(ref _threadState);
```

Comment:
// interlocked read would eliminate the possibility //
It would not.  Consider the following scenario:
* Thread T2 reads _threadState.
* Thread T2 is pre-empted.
* Possibly "much later in time" thread T2 is resumed and now uses a stale _threadState value for behavioral purposes.

An interlocked read may shorten the window, but it cannot eliminate it.